### PR TITLE
RFC-163: Simplify our CDN config, and standardise on Terraform for deployment

### PR DESCRIPTION
**Pull request by robinjam**
[Trello card](https://trello.com/c/3zDgJzZq/3218-write-rfc-proposing-retiring-fastly-vcl)

[✨ Rendered version](https://github.com/alphagov/govuk-rfcs/blob/cdn-rationalisation/rfc-163-cdn-rationalisation.md)

Deadline for comments: **Friday 25th of August**

### CODE REVIEW

File: rfc-163/relocating-logic-from-cdn.md
Comment (ChrisBAshton) [inline rfc-163/relocating-logic-from-cdn.md:8]: I would question whether we _need_ to preserve it. But definitely something to consider (post-RFC):

```suggestion
  - This functionality is currently unused (the dictionary that the denylist is read from is empty), but it exists in case we ever need to quickly block IP addresses (for example, during an incident).
```

File: rfc-163-cdn-rationalisation.md
Comment (ChrisBAshton) [inline rfc-163-cdn-rationalisation.md:38]: I seem to remember reading something along these lines, but I can’t find a web page to back that up - can you provide a link if possible?

File: rfc-163/relocating-logic-from-cdn.md
Comment (ChrisBAshton) [inline rfc-163/relocating-logic-from-cdn.md:37]: Worth linking to https://govuk-kubernetes-cluster-user-docs.publishing.service.gov.uk/manage-app/set-env-var/#update-an-ordinary-non-secret-environment-variable here?

File: rfc-163/relocating-logic-from-cdn.md
Comment (ChrisBAshton) [inline rfc-163/relocating-logic-from-cdn.md:49]: Why only in staging and production?

File: rfc-163/relocating-logic-from-cdn.md
Comment (ChrisBAshton) [inline rfc-163/relocating-logic-from-cdn.md:105]: This was removed in https://github.com/alphagov/govuk-cdn-config/pull/464.

File: rfc-163/relocating-logic-from-cdn.md
Comment (ChrisBAshton) [inline rfc-163/relocating-logic-from-cdn.md:1]: Really good job summarising each VCL feature here - thanks! ⭐

**Review by ChrisBAshton — COMMENTED** [review]
Great work, @robinjam! Some comments below, and I'll leave an aside comment on the Trello card too.

File: rfc-163-cdn-rationalisation.md
Comment (robinjam) [inline rfc-163-cdn-rationalisation.md:79]: It looks like @samsimpson1 from Platform Engineering is already looking into this, but is currently on leave - I will check in when he returns and update this section accordingly

File: rfc-163-cdn-rationalisation.md
Comment (robinjam) [reply rfc-163-cdn-rationalisation.md]: I suppose "their recommended approach" may be an exaggeration - their dev docs homepage points you towards Compute@Edge as opposed to VCL now, but I think both approaches are still supported. Will rephrase this!

File: rfc-163/relocating-logic-from-cdn.md
Comment (sihugh) [inline rfc-163/relocating-logic-from-cdn.md:35]: I think this should be removed. It's certainly not a pattern we want to repeat!

File: rfc-163/relocating-logic-from-cdn.md
Comment (sihugh) [inline rfc-163/relocating-logic-from-cdn.md:48]: Query string allow lists could be a feature in future. Not something to worry about immediately though.

File: rfc-163/relocating-logic-from-cdn.md
Comment (sihugh) [inline rfc-163/relocating-logic-from-cdn.md:27]: I'm conscious that I've recently added this to some of our other CDN services (Bouncer, DGU etc.) Do we propose to move this in all of the services, or just GOV.UK?

File: rfc-163/relocating-logic-from-cdn.md
Comment (sihugh) [inline rfc-163/relocating-logic-from-cdn.md:37]: 🙌

File: rfc-163-cdn-rationalisation.md
Comment (sihugh) [inline rfc-163-cdn-rationalisation.md:5]: Is this scoped just for the GOV.UK service in Fastly, or does it indicate where some of the others should go as well?

File: rfc-163-cdn-rationalisation.md
Comment (sihugh) [inline rfc-163-cdn-rationalisation.md:6]: I think this is applicable across all of the services.

File: rfc-163/relocating-logic-from-cdn.md
Comment (sihugh) [inline rfc-163/relocating-logic-from-cdn.md:13]: We can use custom response bodies in AWS WAF rules 👍 : https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/wafv2_web_acl.html#custom_response_body-block
They're hard coded too, unfortunately. It would be nice to read it from a file on a mirror or something

**Review by sihugh — COMMENTED** [review]
Looks good! Just a few comments

File: rfc-163/relocating-logic-from-cdn.md
Comment (robinjam) [reply rfc-163/relocating-logic-from-cdn.md:26]: Good shout, I haven't looked into the other services (should do so). It might make sense to leave this logic in the VCL then.

Alternatively, we could implement this via WAF for services which have one (www, bouncer, any others?), and leave it in the VCL for the ones that don't.

Not sure which option I prefer, to be honest 🤔

File: rfc-163-cdn-rationalisation.md
Comment (robinjam) [reply rfc-163-cdn-rationalisation.md:5]: Good point! So far I've only looked at the GOV.UK service, but I should probably expand the scope of this RFC to include all of them...

File: rfc-163/relocating-logic-from-cdn.md
Comment (richardTowers) [inline rfc-163/relocating-logic-from-cdn.md:8]: I think there's a problem with these three around caching.

If the CDN has a page in its cache, then it won't make a request to AWS (where the WAF lives), so the IP allowlisting decision / basic auth / denylisting decision happens to late. For example, if a user with a banned IP requests a page that's cached, how is the CDN to know it shouldn't serve that to them?

We could work around this by disabling caching (probably not acceptable, even in non-prod environments as it would be a big disparity between the environments), or maybe by [doing something clever with cookies and the Vary header](https://www.fastly.com/blog/best-practices-using-vary-header#vary-cookie) (but that might end up being more complex / confusing than the current situation).

I think this is worth a bit of a discussion, because it's possibly a showstopper. It's quite possible I've got the wrong end of the stick though.

Comment (robinjam) [reply rfc-163/relocating-logic-from-cdn.md]: @ChrisBAshton and I had a chat about this earlier and we actually think it's probably ok for denylisting to happen in the WAF.

If a denylisted user makes a request of any kind, they will be blocked by the WAF and get an HTTP 403 in response. Origin won't be hit, and client errors aren't (or at least, shouldn't be) cached by the CDN.

If a non-denylisted user makes a cacheable request (e.g. a `GET` to the homepage), they will get a 200 and this will be cached. If the denylisted user then makes the same request, they will receive the cached version (and again, origin won't be hit).

We think that's OK, because to our knowledge denylisting is intended to protect origin from attacks, rather than for access control.

Basic auth will still have to live in the CDN for the reason you mentioned (otherwise we'd have to add `Vary: Authorization` to the response, which would pretty much defeat caching).

What do you think @richardTowers ?

Comment (richardTowers) [reply rfc-163/relocating-logic-from-cdn.md]: > denylisting is intended to protect origin from attacks, rather than for access control

It's true that it's not used for access control, but I'm not sure if that means it's just there to protect origin. It also gives us some useful protection at the CDN level. For example, if someone's downloading terrabytes of assets from one IP, Fastly might ask us if we could make them stop doing that (or if Fastly ever move us to usage based billing, we might want to make them stop), and I think IP denylisting is the only way we could do that at the moment.

> client errors aren't (or at least, shouldn't be) cached by the CDN

Definitely [citation needed] for this - I seem to remember 404s being cached (which is a reasonable choice for a CDN to make - there wasn't a resource here last time it was requested, there won't be a resource here this time). I'm not sure whether 403s get different treatment. We could fix that in the VCL for sure, if we haven't already (but then we're adding more code to the VCL to enable us to remove some code from the VCL, so we'd have to consider the cost / benefit of that).

> otherwise we'd have to add Vary: Authorization to the response, which would pretty much defeat caching

🤔  I wonder how bad this would be actually. Browsers don't do any kind of nonce / salting of Authorization headers, so they're always the same for the same username / password. Our basic auth use cases don't have a massive variety of username / password combinations, so I think we'd get a reasonable cache hit rate for them. Maybe that one could move to the WAF.

Comment (richardTowers) [reply rfc-163/relocating-logic-from-cdn.md]: https://www.rfc-editor.org/rfc/rfc7231#section-6.5.4 lists a few 4xx responses as cacheable by default, but notably not 403s (404s are cacheable, as are 405s, 410s, and a few others). Whether Varnish implements this is another question...

Comment (richardTowers) [reply rfc-163/relocating-logic-from-cdn.md]: https://github.com/varnish/Varnish-Book/blob/master/varnish_book.rst#the-initial-value-of-beresp-ttl

> Only the following status codes will be cached by default:

> 200: OK
> 203: Non-Authoritative Information
> 300: Multiple Choices
> 301: Moved Permanently
> 302: Moved Temporarily
> 304: Not modified
> 307: Temporary Redirect
> 410: Gone
> 404: Not Found

So not quite the same as RFC7231, but it won't cache 403s so you can ignore that argument :)

Comment (robinjam) [reply rfc-163/relocating-logic-from-cdn.md]: > For example, if someone's downloading terrabytes of assets from one IP, Fastly might ask us if we could make them stop doing that (or if Fastly ever move us to usage based billing, we might want to make them stop), and I think IP denylisting is the only way we could do that at the moment.

I think we'd be more likely to use JA3 denylisting for this(?)

I'm actually not sure how valuable IP denylisting is as a general concept, given how easy it is for an attacker to get access to a bunch of IPv4 addresses, and given that IPv6 is a thing, but I'm open to being persuaded 😅

> > otherwise we'd have to add Vary: Authorization to the response, which would pretty much defeat caching
> 
> 🤔 I wonder how bad this would be actually. Browsers don't do any kind of nonce / salting of Authorization headers, so they're always the same for the same username / password. Our basic auth use cases don't have a massive variety of username / password combinations, so I think we'd get a reasonable cache hit rate for them. Maybe that one could move to the WAF.

Unfortunately I think it'd have to be more complicated than that, because we allow access to integration if _either_ the user provided the password _or_ their IP is in the allowlist. We'd have to do something like either:

- Set `Vary: Fastly-Client-IP, Authorization` on the response (which would essentially defeat caching), or
- Set `Vary: Authorization` always, and `Vary: Fastly-Client-IP` only if the `Authorization` header is missing (which actually might be a good shout, depending on how complex it'd be to implement in WAF 🤔)

I think I might tentatively move that back to the "move to WAF" section, with a caveat that it might prove to be too complicated.

Comment (robinjam) [reply rfc-163/relocating-logic-from-cdn.md]: Thanks, that's useful context! I think we'll need to leave the denylist in the CDN, in that case (and see what we can do about ensuring parity between Fastly and Cloudfront)

File: rfc-163/relocating-logic-from-cdn.md
Comment (theseanything) [inline rfc-163/relocating-logic-from-cdn.md:9]: Could this even be done at the application level? I think there's potential we create ourselves another problem of being locked into using AWS WAF. Which would be an issue if we wanted to look at an alt cloud provider - (for example Disaster Recovery)

Comment (robinjam) [reply rfc-163/relocating-logic-from-cdn.md:9]: That's a good point actually, I hadn't considered that we'd be moving essentially from one form of vendor lock-in to another.

My concern with e.g. the `security.txt` redirects is that we have the same functionality set up across several different domains (including gov.uk, assets.gov.uk, service.gov.uk and DGU). If we want to move this logic to the application layer then it'll presumably need to be implemented in a different way for each domain, whereas if we move it to the WAF then the implementation is consistent for each.

File: rfc-163-cdn-rationalisation.md
Comment (ChrisBAshton) [inline rfc-163-cdn-rationalisation.md:41]: Good suggestion 👍

File: rfc-163/relocating-logic-from-cdn.md
Comment (ChrisBAshton) [reply rfc-163/relocating-logic-from-cdn.md:9]: I'd be inclined to do the security.txt redirect in Router, and consider the redirect on service.gov.uk and DGU to be out of scope for this RFC, as they're already not working:

- https://service.gov.uk/security.txt ->

File: rfc-163-cdn-rationalisation.md
Comment (ChrisBAshton) [inline rfc-163-cdn-rationalisation.md:34]: I think it would be good to hit home that we don't want more VCL being written:

```suggestion

### Avoid adding new VCL unless absolutely necessary

See section above. Wherever possible, we should consider where else our logic might live, whether that's at the application layer, the Router layer or the WAF layer. Now that we've started using Terraform to manage our Fastly configuration, some features may also be able to be represented in Terraform rather than VCL.
```

**Review by ChrisBAshton — COMMENTED** [review]
Hi @robinjam, apologies this took me longer than I'd have liked to get to!

I think we should get this merged in soon, but it seems a shame to merge it when a fair chunk of it is now somewhat outdated (the nature of longstanding RFCs!). I've added some comments below on areas I think would be good to update before merge. Thank you ⭐

**Review by sihugh — APPROVED** [review]

File: rfc-163/relocating-logic-from-cdn.md
Comment (robinjam) [reply rfc-163/relocating-logic-from-cdn.md:32]: I think I'd rather leave it here, because that card was supposed to be an outcome of this RFC (they ended up getting delivered out of order because this RFC was held up waiting for approval 😅)

I also want to keep the context around feature flags in general

File: rfc-163/relocating-logic-from-cdn.md
Comment (robinjam) [reply rfc-163/relocating-logic-from-cdn.md:64]: I'll make a note to investigate this further (or write up a card for it) 👍

**Comment by robinjam** [comment]
Thanks very much for the feedback everyone 🙏

File: rfc-163/relocating-logic-from-cdn.md
Comment (ChrisBAshton) [reply rfc-163/relocating-logic-from-cdn.md:36]: Sounds good 👍

**Pull request merged**
